### ISSUE TITLE
Opening subtractions fail on wall (example file)

### ISSUE THREAD

**Issue by johanrd** [IfcOpenShell/IfcOpenShell#992]
1. Convert [992--wall--missingSubtractions.ifc](https://github.com/IfcOpenShell/files/blob/master/992--wall--missingSubtractions.ifc) with 55a54098dd1974dcfb5d8bf6c6c249886a528092 (OCCT 7.3.0p4) and experience failing opening subtractions:

<img width="300" alt="Screenshot 2020-09-08 at 23 53 35" src="https://user-images.githubusercontent.com/4601554/92531457-93c80c80-f22e-11ea-8fab-103c07e1ff5d.png">

```
#1157=IfcWallStandardCase('2bzAeH_xr7NPJBIGOb5PFj',#1,'wall',$,'Basic Wall:XIVB-TEGEL-575',#2295,#691170,'4059236')
[Notice] {2bzAeH_xr7NPJBIGOb5PFj} Boolean operation yields non-manifold result
[Notice] {2bzAeH_xr7NPJBIGOb5PFj} Boolean operation yields non-manifold result
[Notice] {2bzAeH_xr7NPJBIGOb5PFj} Boolean operation yields non-manifold result
[Notice] {2bzAeH_xr7NPJBIGOb5PFj} No longer attempting boolean operation with higher fuzziness
[Error] {2bzAeH_xr7NPJBIGOb5PFj} Opening subtraction failed for 10 openings
```

2. Expected geometry from Solibri:

<img width="300" alt="Screenshot 2020-09-08 at 23 52 30" src="https://user-images.githubusercontent.com/4601554/92531486-a3475580-f22e-11ea-9fe9-0f47a35f83b4.png">

**Comment by chingham** [comment]
Did you manage to solve this issue ? I have the exact same one

**Comment by johanrd** [comment]
@chingham There are many occasions for why/how an opening subtraction may fail. I have the impresson that this most often happens when the coordinates of the opening subtraction are just slightly off from the wall it should be subtracted from, but there are almost endless sources of failure of boolean operations.

[IfcConvert is not yet as robust as we could hope for](https://github.com/IfcOpenShell/IfcOpenShell/issues?q=is%3Aissue+subtractions+wall+) regarding opening subtractions. I would encourage you to create a separate issue with your wall isolated – that will make it easier to debug and verify the robustness when improving the subtraction handling. 

For debugging, and in a hope to find a temporary solution, I can recommend to build IfcConvert with other versions of OCCT (e.g. 7.3.0, 7.3.0p4 and 7.4.0). And/or try the pre-built versions of IfcConvert, e.g. [8625aab](https://github.com/IfcOpenBot/IfcOpenShell/commit/8625aab5f0c2c4752eb10ad20daee744547b377e#comments) and [bf16239](https://github.com/IfcOpenBot/IfcOpenShell/commit/bf1623975c6451baf3e2a1af9f32e9306d235b3f#comments).

There are hopes that boolean operations will be more easy to work with in the 0.7-branch where CGAL is added as an alternative to OCCT. @aothms may want to chip in with his thoughts on this.

**Comment by aothms** [comment]
@johanrd is correct.

 In this particular case the issue is actually fully caused by IfcOpenShell. There are boolean operands that share an edge:

![afbeelding](https://user-images.githubusercontent.com/1096535/98011808-a1e27580-1df8-11eb-9e2f-32f6f74e464f.png)

![afbeelding](https://user-images.githubusercontent.com/1096535/98011824-a60e9300-1df8-11eb-801b-e5bd6abe9755.png)

Causing the result to be non-manifold which IfcOpenShell does not allow (because more often than not it means an error in processing). I have not found a way yet to distinguish valid non-manifold results from invalid ones.


We have implemented an additional work around for this but it doesn't work for cylindrical operands.

![afbeelding](https://user-images.githubusercontent.com/1096535/98013084-313c5880-1dfa-11eb-9019-39b3acd57dd6.png)

In case of Openings A and B we would detect there is an overlapping edge from faces that that do not overlap. But in case of a cylinder in the OCCT Brep model an edge is only actually composed of three edges. So overlapping edges are never detected. So we need to generalize this workaround to non-polyhedral operands that are touching.

**Comment by BIMvoice** [comment]
This is the coincident-tool-face boolean case addressed by #8489. Each of the 10 openings is an extrusion whose depth equals the wall thickness (575), so the opening end caps land exactly coincident with the wall's two large faces, and the cut returns a non-manifold result that is rejected at every fuzziness, dropping all openings (the wall renders uncut).

Verified locally against v0.8.0 + #8489 (OCC 7.9.2): with that change the openings are cut. Wall 2bzAeH_xr7NPJBIGOb5PFj volume drops from 21.998 to 14.220, the ~7.78 removed matching the 10 openings (dominated by the large 3500x3850 opening). Should be safe to close once #8489 lands.
